java/ql/src/experimental/Security/CWE/CWE-522/InsecureBasicAuth.ql

```diff
@@ -32,21 +32,47 @@ class ApacheHttpRequest extends RefType {
 class URLConstructor extends ClassInstanceExpr {
   URLConstructor() { this.getConstructor().getDeclaringType() instanceof TypeUrl }
 
-  Expr stringArg() {
-    // URLs constructed with any of the four string constructors below:
-    // URL(String spec), URL(String protocol, String host, int port, String file), URL(String protocol, String host, int port, String file, URLStreamHandler handler), URL(String protocol, String host, String file)
+  // URLs constructed with the string constructor `URL(String spec)`
+  Expr specArg() {
     this.getConstructor().getParameter(0).getType() instanceof TypeString and
-    result = this.getArgument(0) // First argument contains the protocol part.
+    (
+      this.getConstructor().getNumberOfParameters() = 1 and
+      result = this.getArgument(0) // First argument contains the whole spec.
+    )
+  }
+
+  // URLs constructed with any of the three string constructors below:
+  // `URL(String protocol, String host, int port, String file)`,
+  // `URL(String protocol, String host, int port, String file, URLStreamHandler handler)`,
+  // `URL(String protocol, String host, String file)`
+  Expr protocolArg() {
+    this.getConstructor().getParameter(0).getType() instanceof TypeString and
+    (
+      this.getConstructor().getNumberOfParameters() > 1 and
+      result = this.getArgument(0) // First argument contains the protocol part.
+    )
+  }
+
+  Expr hostArg() {
+    this.getConstructor().getParameter(0).getType() instanceof TypeString and
+    (
+      this.getConstructor().getNumberOfParameters() > 1 and
+      result = this.getArgument(1) // Second argument contains the host part.
+    )
   }
 }
 
 /**
- * The type `java.net.URLConnection`.
+ * Gets a regular expression for matching private hosts.
  */
-class TypeHttpUrlConnection extends RefType {
-  TypeHttpUrlConnection() {
-    this.getASourceSupertype*().hasQualifiedName("java.net", "HttpURLConnection")
-  }
+private string getPrivateHostRegex() {
+  result = "localhost(/.*)?" or
```

Comment:
It appears host names are case-insensitive (see [RFC 4343](https://tools.ietf.org/html/rfc4343)), therefore `localhost` should probably be matched case insensitively:
```suggestion
  result = "(?i)localhost(/.*)?" or
```

Reply:
Thanks @Marcono1234 for the advice. I've made the change.